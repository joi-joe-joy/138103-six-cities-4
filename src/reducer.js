import {extend} from "./utils.js";
import offers from "./mocks/offers.js";

let initCity = offers[0].city;
let initOffers = offers.filter((offer) => offer.city === initCity);

// Получаем все значения городов
let cities = offers.map((offer)=>offer.city);
// Получаем уникальные значения городов
let citiesList = Array.from(new Set(cities));

const initialState = {
  city: initCity,
  offersCityList: initOffers,
  offers,
  citiesList
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  GET_OFFERS_CITY_LIST: `GET_OFFERS_CITY_LIST`,
};

const ActionCreator = {
  changeCity: (city) => {
    let changedCity = city ? city : offers[0].city;

    return {
      type: ActionType.CHANGE_CITY,
      payload: changedCity
    };
  },
  getOffersCityList: (city) => {
    let offersCityList = offers.filter((offer) => offer.city === city);

    return {
      type: ActionType.GET_OFFERS_CITY_LIST,
      payload: offersCityList
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        city: action.payload
      });
    case ActionType.GET_OFFERS_CITY_LIST:
      return extend(state, {
        offersCityList: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator};
