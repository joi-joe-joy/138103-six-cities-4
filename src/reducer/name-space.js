export const NameSpace = {
  DATA: `DATA`,
  USER: `USER`,
  PLACE: `PLACE`
};
