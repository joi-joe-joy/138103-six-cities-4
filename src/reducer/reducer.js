import {combineReducers} from "redux";
import {reducer as data} from "./data/data";
import {reducer as place} from "./place/place";
import {reducer as user} from "./user/user";
import {reducer as comments} from "./comments/comments";
import {reducer as favorite} from "./favorite/favorite";
import {NameSpace} from "./name-space";

export default combineReducers({
  [NameSpace.DATA]: data,
  [NameSpace.PLACE]: place,
  [NameSpace.USER]: user,
  [NameSpace.COMMENTS]: comments,
  [NameSpace.FAVORITE]: favorite,
});
