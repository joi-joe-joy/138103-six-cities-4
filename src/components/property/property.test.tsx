import * as React from "react";
import * as renderer from "react-test-renderer";
import Property from "./property";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {NameSpace} from "../../reducer/name-space";

const mockStore = configureStore([]);

const offerWithPremium = {
  bedrooms: 2,
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  goods: [`Air conditioning`, `Laptop friendly workspace`, `Baby seat`, `Fridge`, `Breakfast`, `Washer`, `Washing machine`, `Dishwasher`, `Coffee machine`, `Towels`],
  host: {id: 25, name: `Angelina`, isPro: true, avatarUrl: `img/avatar-angelina.jpg`},
  id: 1,
  images: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`],
  isFavorite: false,
  isPremium: true,
  location: {
    latitude: 48.865610000000004,
    longitude: 2.350499,
    zoom: 16
  },
  maxAdults: 8,
  previewImage: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`,
  price: 397,
  rating: 3.6,
  title: `Penthouse, 4-5 rooms + 5 balconies`,
  type: `hotel`
};

const offerWithoutPremium = {
  bedrooms: 2,
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  goods: [`Air conditioning`, `Laptop friendly workspace`, `Baby seat`, `Fridge`, `Breakfast`, `Washer`, `Washing machine`, `Dishwasher`, `Coffee machine`, `Towels`],
  host: {id: 25, name: `Angelina`, isPro: true, avatarUrl: `img/avatar-angelina.jpg`},
  id: 1,
  images: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`],
  isFavorite: false,
  isPremium: true,
  location: {
    latitude: 48.865610000000004,
    longitude: 2.350499,
    zoom: 16
  },
  maxAdults: 8,
  previewImage: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`,
  price: 397,
  rating: 3.6,
  title: `Penthouse, 4-5 rooms + 5 balconies`,
  type: `hotel`
};

const comments = [{
  comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
  date: `2019-05-08T14:13:56.569Z`,
  id: 1,
  rating: 4,
  user: {
    avatarUrl: `img/1.png`,
    id: 4,
    isPro: false,
    name: `Max`
  }
}];

it(`Render Property correctly`, () => {
  const store = mockStore({
    [NameSpace.PLACE]: {
      currentCard: {}
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`
    },
    [NameSpace.DATA]: {
      city: {
        name: `Paris`,
        location: {
          latitude: 48.85661,
          longitude: 2.351499,
          zoom: 13
        }
      },
      offers: [offerWithPremium],
      offersNearby: [offerWithPremium],
    },
    [NameSpace.COMMENTS]: {
      comments
    },
  });

  const tree = renderer.create(
      <Provider store={store}>
        <Property
          authStatus={`AUTH`}
          city={{
            name: `Paris`,
            location: {
              latitude: 48.85661,
              longitude: 2.351499,
              zoom: 13
            }
          }}
          offer={offerWithPremium}
          nearbyOffers={[offerWithPremium]}
        />
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Property render correctly without Premium`, () => {
  const store = mockStore({
    [NameSpace.PLACE]: {
      currentCard: {}
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`
    },
    [NameSpace.DATA]: {
      city: {
        name: `Paris`,
        location: {
          latitude: 48.85661,
          longitude: 2.351499,
          zoom: 13
        }
      },
      offers: [offerWithPremium],
      offersNearby: [offerWithPremium],
    },
    [NameSpace.COMMENTS]: {
      comments
    },
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Property
            authStatus={`AUTH`}
            city={{
              name: `Paris`,
              location: {
                latitude: 48.85661,
                longitude: 2.351499,
                zoom: 13
              }
            }}
            offer={offerWithoutPremium}
            nearbyOffers={[offerWithPremium]}
          />
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
