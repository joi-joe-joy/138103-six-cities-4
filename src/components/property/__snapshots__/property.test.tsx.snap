// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property render correctly without Premium 1`] = `null`;

exports[`Render Property correctly 1`] = `null`;
